experimental/ir/ir_symbol.go

```diff
@@ -213,3 +214,168 @@ func (s symtab) sort(c *Context) {
+//
+// If skipIfNot is nil, the symbol's kind will not be checked to determine if
+// we should continue climbing scopes.
+func (s symtab) resolve(
+	c *Context,
+	scope, name FullName,
+	skipIfNot func(SymbolKind) bool,
+) (found ref[rawSymbol], expected FullName) {
+	// This function implements the name resolution algorithm specified at
+	// https://protobuf.com/docs/language-spec#reference-resolution.
+
+	// Symbol resolution is not quite as simple as trying p + name for all
+	// ancestors of scope. Consider the following files:
+	//
+	// 	// a.proto
+	// 	package foo.bar;
+	// 	message M {}
+	//
+	// 	// b.proto
+	//  package foo;
+	//  import "a.proto";
+	//  message M {}
+	//
+	//  // c.proto
+	//  package foo.bar.baz;
+	//  import "b.proto";
+	//  message N {
+	//    M m = 1;
+	//  }
+	//
+	// The candidates, in order, are:
+	// - foo.bar.baz.M; does not exist.
+	// - foo.bar.M; not visible.
+	// - foo.M; correct answer.
+	// - M; not tried.
+	//
+	// If we do not keep going after encountering symbols that are not visible
+	// to us, we will reject valid code.
+
+	// A similar situation happens here:
+	//
+	// package foo;
+	// message M {
+	//   message N {}
+	//   message P {
+	//     enum X { N = 1; }
+	//     N n = 1;
+	//   }
+	// }
+	//
+	// If we look up N, the candidates are foo.M.P.N, foo.M.N, foo.N, and N.
+	// We will find foo.M.P.N, which is not a message or enum type, so we must
+	// skip it to find the correct name, foo.M.N. This is what the accept
+	// predicate is for.
+
+	// Finally, consider the following situation, which involves partial
+	// names.
+	//
+	// package foo;
+	// message M {
+	//   message N {}
+	//   message M {
+	//     M.N n = 1;
+	//   }
+	// }
+	//
+	// The candidates are foo.M.M.N, foo.M.N, M.N. However, protoc rejects this,
+	// because it actually searches for M first, and then appends the rest of
+	// the path and searches for that, in two phases.
+	//
+	// It is not clear why protoc does this, but it does mean we need to be
+	// careful in how we resolve partial names.
+
+	scopeSearch := !name.IsIdent()
+	first := name.First()
+	candidate := []byte(scope.Append(first))
+
+	// Adapt accept to account for scopeSearch and to be ok to call if nil.
+	accept := func(kind SymbolKind) bool {
+		if scopeSearch {
+			return kind.IsScope()
+		}
+		return skipIfNot == nil || skipIfNot(kind)
+	}
+
+again:
+	for {
+		ref := s.lookupBytes(c, candidate)
+		if !ref.ptr.Nil() {
+			found = ref
+			sym := wrapSymbol(c, ref)
+			if sym.Visible() && accept(sym.Kind()) {
+				// If the symbol is not visible, keep looking; we may find
+				// another match that is actually visible.
+				break
+			}
+		}
+
+		if scope == "" {
+			// Out of places to look. This is probably a fail.
+			break
+		}
+		oldLen := len(scope)
+		scope = scope.Parent()
+		if scope == "" {
+			oldLen++
+		}
+		// Delete in-place to avoid spamming allocations for each candidate.
+		candidate = slices.Delete(candidate, len(scope), oldLen)
```

Comment:
I thought the main value of using `slices.Delete` for something like this, instead of the simpler `candidate[:len(scope)]`, what that it cleared slice elements to zero values (like to avoid having no-longer-used slots in the backing array continue to pin values to the heap). But since this is a `[]byte`, there's no need to clear elements. 🤔

Reply:
This is deleting from the middle of the slice, so e.g. `foo.bar.Baz` -> `foo.Baz`. We unfortunately need to materialize the whole string because Go won't let us look up strings in a map using only parts of a string.